plugins/fabric8-cdn-support/src/main/java/com/redhat/che/cdn/CdnSupportService.java

```diff
@@ -0,0 +1,326 @@
+      @Nullable @Named("che.fabric8.cdn.prefetch.editor") String editorToPrefetch) {
+    this(RUNNER, pluginMetaRetriever, editorToPrefetch);
+  }
+
+  @VisibleForTesting
+  CdnSupportService(
+      CommandRunner commandRunner,
+      PluginMetaRetriever pluginMetaRetriever,
+      @Nullable @Named("che.fabric8.cdn.prefetch.editor") String editorToPrefetch) {
+    this.pluginMetaRetriever = pluginMetaRetriever;
+    this.editorToPrefetch = editorToPrefetch;
+    this.commandRunner = commandRunner;
+
+    // Test that the skopeo process is available
+
+    try {
+      commandRunner.runCommand(
+          SKOPEO_BINARY,
+          SKOPEO_HELP_ARGS,
+          null,
+          SKOPEO_HELP_TIMEOUT_SECONDS,
+          TimeUnit.SECONDS,
+          null,
+          null);
+    } catch (IOException e) {
+      throw new RuntimeException(
+          "The `skopeo` command is not available. Please check that is has been correctly installed in the Che server Docker image",
+          e);
+    } catch (ExecutionException | InterruptedException | TimeoutException e) {
+      LOG.warn("Exception during the `skopeo --help` command execution", e);
+    }
+  }
+
+  @GET
+  @Path("paths")
+  @Produces(APPLICATION_JSON)
+  public String getPaths(@Context HttpServletRequest servletRequest) throws Exception {
+    if (editorToPrefetch == null) {
+      throw new NotFoundException("No editor is configured for CDN resource pre-fetching");
+    }
+
+    LOG.debug("Searching the editor plugin entry in the plugin registry");
+
+    Collection<PluginMeta> plugins =
+        pluginMetaRetriever.get(
+            ImmutableMap.of(Constants.WORKSPACE_TOOLING_EDITOR_ATTRIBUTE, editorToPrefetch));
+    if (plugins.size() != 1) {
+      throw new NotFoundException("Editor '" + editorToPrefetch + "' is unknown");
+    }
+
+    PluginMeta editor = plugins.toArray(new PluginMeta[] {})[0];
+    if (editor == null) {
+      throw new NotFoundException("Editor '" + editorToPrefetch + "' is unknown");
+    }
+
+    LOG.debug("Retrieving editor URL");
+
+    String url = editor.getUrl();
+    if (url == null) {
+      throw new BadRequestException(format("URL of editor '%s' is null", editorToPrefetch));
+    }
+    if (!url.equals(editorDefinitionUrl)) {
+      editorDefinitionUrl = url;
+      dockerImage = null;
+    } else {
+      LOG.debug("Editor full definition didn't change");
+    }
+
+    if (dockerImage == null) {
+      dockerImage = readDockerImageName(editorDefinitionUrl);
+    }
+
+    if (dockerImage == null) {
+      throw new BadRequestException(
+          format(
+              "Plugin container image not found in the plugin descriptor of '%s'",
+              editorToPrefetch));
+    }
+
+    LOG.debug("Running {} on image {}", SKOPEO_BINARY, dockerImage);
+    final ListLineConsumer out = commandRunner.newOutputConsumer();
+    final ListLineConsumer err = commandRunner.newErrorConsumer();
+    Process skopeoInspect =
```

Comment:
Having all the business logic and all the levels of abstraction in a single place would make maintenance of the code harder. Consider refactoring the code. SOLID is what inspires me when I struggle on structuring the code.

Reply:
I refactored a bit to have the main `getPaths` function simpler and clearer